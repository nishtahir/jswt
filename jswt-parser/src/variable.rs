use jswt_ast::{Annotation, VariableDeclarationElement};
use jswt_common::{Span, Spannable};
use jswt_tokenizer::TokenType;

use crate::{consume, ident, Identifier, ParseError, ParseResult, Parser};

impl<'a> Parser<'a> {
    ///VariableDeclaration
    ///  :  VariableModifier? Identifier TypeAnnotation? '=' SingleExpression ';'
    ///  ;
    pub(crate) fn variable_declaration(
        &mut self,
        annotations: Vec<Annotation>,
        export: bool,
    ) -> ParseResult<VariableDeclarationElement> {
        let modifier = self.variable_modifier()?;
        let name = ident!(self)?;

        let mut type_annotation = None;
        if self.lookahead_is(TokenType::Colon) {
            type_annotation = Some(self.type_annotation()?);
        }

        consume!(self, TokenType::Equal)?;

        let expression = self.single_expression()?;
        let end = consume!(self, TokenType::Semi)?;

        Ok(VariableDeclarationElement {
            annotations,
            export,
            span: modifier.span() + end,
            modifier,
            name,
            expression,
            type_annotation,
        })
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use jswt_assert::assert_debug_snapshot;
    use jswt_tokenizer::Tokenizer;

    #[test]
    fn test_basic_variable_declaration() {
        let mut tokenizer = Tokenizer::default();
        tokenizer.enqueue_source_str("test_basic_variable_declaration", "const x = 1;");
        let mut parser = Parser::new(&mut tokenizer);
        let actual = parser.parse();
        assert_debug_snapshot!(actual);
    }

    #[test]
    fn test_variable_declaration_with_type_annotation() {
        let mut tokenizer = Tokenizer::default();
        tokenizer.enqueue_source_str(
            "test_variable_declaration_with_type_annotation",
            "@Test const x: i32 = 1;",
        );
        let mut parser = Parser::new(&mut tokenizer);
        let actual = parser.parse();
        assert_debug_snapshot!(actual);
    }

    #[test]
    fn test_variable_declaration_with_export() {
        let mut tokenizer = Tokenizer::default();
        tokenizer.enqueue_source_str("test_variable_declaration_with_export", "export const x = 1;");
        let mut parser = Parser::new(&mut tokenizer);
        let actual = parser.parse();
        assert_debug_snapshot!(actual);
    }
}
