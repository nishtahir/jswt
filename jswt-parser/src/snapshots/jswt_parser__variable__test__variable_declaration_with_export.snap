---
source: jswt-parser/src/variable.rs
expression: actual

---
Ast {
    program: Program {
        files: [
            File {
                span: Span {
                    file: "test_variable_declaration_with_export",
                    module: "module/test_variable_declaration_with_export",
                    start: 0,
                    end: 19,
                },
                source_elements: SourceElements {
                    span: Span {
                        file: "test_variable_declaration_with_export",
                        module: "module/test_variable_declaration_with_export",
                        start: 0,
                        end: 19,
                    },
                    source_elements: [
                        VariableDeclaration(
                            VariableDeclarationElement {
                                annotations: [],
                                export: true,
                                span: Span {
                                    file: "test_variable_declaration_with_export",
                                    module: "module/test_variable_declaration_with_export",
                                    start: 7,
                                    end: 19,
                                },
                                modifier: Const(
                                    Span {
                                        file: "test_variable_declaration_with_export",
                                        module: "module/test_variable_declaration_with_export",
                                        start: 7,
                                        end: 12,
                                    },
                                ),
                                name: Identifier {
                                    span: Span {
                                        file: "test_variable_declaration_with_export",
                                        module: "module/test_variable_declaration_with_export",
                                        start: 13,
                                        end: 14,
                                    },
                                    value: "x",
                                },
                                expression: Literal(
                                    Integer(
                                        IntegerLiteral {
                                            span: Span {
                                                file: "test_variable_declaration_with_export",
                                                module: "module/test_variable_declaration_with_export",
                                                start: 17,
                                                end: 18,
                                            },
                                            value: 1,
                                            ty: Unknown,
                                        },
                                    ),
                                ),
                                type_annotation: None,
                            },
                        ),
                    ],
                },
            },
        ],
    },
}
